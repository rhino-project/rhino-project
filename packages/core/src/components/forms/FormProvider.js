import PropTypes from 'prop-types';
import { FormProvider as RHFFormProvider } from 'react-hook-form';
import { DevTool } from '@hookform/devtools';

const FormProvider = ({ children, ...props }) => {
  const { control } = props;

  return (
    <>
      {import.meta.env.DEV && (
        <DevTool
          control={control}
          placement={'top-left'}
          styles={{
            button: {
              width: '20px'
            }
          }}
        />
      )}
      <RHFFormProvider {...props}>{children}</RHFFormProvider>
    </>
  );
};

FormProvider.propTypes = {
  children: PropTypes.node
};

export default FormProvider;
