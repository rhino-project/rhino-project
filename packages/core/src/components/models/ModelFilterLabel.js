import { useMemo } from 'react';
import { useModelFilterField } from 'rhino/hooks/form';
import { useGlobalOverrides } from 'rhino/hooks/overrides';
import FilterLabel from '../forms/FilterLabel';

const isDateRelated = (format) => ['date', 'time', 'datetime'].includes(format);

const operatorToDateLabel = (operator) => {
  switch (operator) {
    case 'diff':
      return 'not';
    case 'gt':
    case 'gteq':
      return 'after';
    case 'lt':
    case 'lteq':
      return 'before';
    default:
      return '';
  }
};

const operatorToLabel = (format, operator) => {
  if (isDateRelated(format)) return operatorToDateLabel(operator);

  switch (operator) {
    case 'diff':
    case 'gt':
      return '>';
    case 'gteq':
      return '>=';
    case 'lt':
      return '<';
    case 'lteq':
      return '<=';
    default:
      return '';
  }
};

const defaultComponents = {
  ModelFilterLabel: FilterLabel
};

export const ModelFilterLabel = ({ overrides, label, model, ...props }) => {
  const { ModelFilterLabel } = useGlobalOverrides(defaultComponents, overrides);

  const { path } = props;
  const { attribute, operator } = useModelFilterField(model, path);

  const modelLabel = useMemo(
    () =>
      label ||
      `${attribute.readableName} ${operatorToLabel(
        attribute.format,
        operator
      )}`,
    [attribute, operator, label]
  );

  return <ModelFilterLabel label={modelLabel} {...props} />;
};

export default ModelFilterLabel;
