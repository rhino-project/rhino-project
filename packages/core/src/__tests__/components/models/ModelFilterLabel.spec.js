import { render } from '@testing-library/react';
import { useForm } from 'react-hook-form';
import FormProvider from 'rhino/components/forms/FormProvider';
import ModelFilterLabel from 'rhino/components/models/ModelFilterLabel';
import { sharedModelTests } from './sharedModelTests';

describe('ModelFilterLabel', () => {
  const wrapper = ({ children }) => {
    const methods = useForm();
    return <FormProvider {...methods}>{children}</FormProvider>;
  };

  sharedModelTests(ModelFilterLabel);

  it(`should render label with readable name`, () => {
    const { asFragment } = render(
      <ModelFilterLabel
        model={{
          properties: { dummy: { type: 'string', readableName: 'Dummy' } }
        }}
        path="dummy"
      />,
      {
        wrapper
      }
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it(`should override label with prop`, () => {
    const { asFragment } = render(
      <ModelFilterLabel
        model={{
          properties: { dummy: { type: 'string', readableName: 'Dummy' } }
        }}
        label="Dummy Override"
        path="dummy"
      />,
      {
        wrapper
      }
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
