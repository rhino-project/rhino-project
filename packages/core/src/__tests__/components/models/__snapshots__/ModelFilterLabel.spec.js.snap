// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelFilterLabel should override label with prop 1`] = `
<DocumentFragment>
  <label
    class="form-label"
    for="dummy"
    id="filterLabel-dummy"
    label="Dummy Override"
    path="dummy"
  >
    Dummy Override
  </label>
</DocumentFragment>
`;

exports[`ModelFilterLabel should render label with readable name 1`] = `
<DocumentFragment>
  <label
    class="form-label"
    for="dummy"
    id="filterLabel-dummy"
    label="Dummy "
    path="dummy"
  >
    Dummy 
  </label>
</DocumentFragment>
`;
